src/mono/mono/mini/aot-compiler.c

```diff
@@ -14355,6 +14387,70 @@ add_interp_in_wrapper_for_sig (MonoAotCompile *acfg, MonoMethodSignature *sig)
 	add_method (acfg, wrapper);
 }
 
+static gboolean
+process_direct_pinvokes (MonoAotCompile *acfg, char *dpi)
+{
+	gboolean processed = TRUE;
+	char *direct_pinvoke = g_strdup (dpi);
+	if (!direct_pinvoke)
+		goto early_exit;
+
+	if (direct_pinvoke[0] == '#')
+		goto early_exit;
+
+	if (!g_strstrip (direct_pinvoke))
+		goto early_exit;
+
+	char **direct_pinvoke_split = g_strsplit (direct_pinvoke, "!", -1);
+	if (!direct_pinvoke_split) {
+		processed = FALSE;
+		aot_printerrf (acfg, "Failed to split the provided 'direct_pinvoke' AOT option '%s' with delimiter '!'\n", dpi);
+		goto cleanup;
+	}
+
+	char *library_name = g_strdup (direct_pinvoke_split[0]);
+	if (!library_name) {
+		processed = FALSE;
+		aot_printerrf (acfg, "Failed to strdup the module '%s' for the provided 'direct_pinvoke' AOT option '%s'.\n", direct_pinvoke_split[0], dpi);
+		goto cleanup;
+	}
+
+	if (!direct_pinvoke_split[1]) {
+		g_hash_table_insert (acfg->direct_pinvokes, library_name, NULL);
+		goto cleanup;
+	}
+
+	if (direct_pinvoke_split[2]) {
+		aot_printerrf (acfg, "The provided 'direct_pinvoke' AOT option '%s' shouldn't contain multiple '!'\n", dpi);
+		processed = FALSE;
+		goto cleanup;
+	}
+
+	char *entrypoint_name = g_strdup (direct_pinvoke_split[1]);
+	if (!entrypoint_name) {
+		processed = FALSE;
+		aot_printerrf (acfg, "Failed to strdup the entrypoint '%s' for the provided 'direct_pinvoke' AOT option '%s'.\n", direct_pinvoke_split[1], dpi);
+		goto cleanup;
+	}
+
+	GHashTable *val;
+	if (g_hash_table_lookup_extended (acfg->direct_pinvokes, library_name, NULL, (gpointer *)&val)) {
+		if (!val || g_hash_table_contains (val, entrypoint_name))
+			goto cleanup;
+		g_hash_table_insert (val, entrypoint_name, NULL);
+	} else {
+		val = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, NULL);
+		g_hash_table_insert (val, entrypoint_name, NULL);
+		g_hash_table_insert (acfg->direct_pinvokes, library_name, val);
+	}
+
+cleanup:
```

Comment:
You still need to handle memory ownership and free's in error cases for direct_pinvoke, library_name, entrypoint_name, alternative is to keep strings in direct_pinvoke_slit until added into the hash table and you do strdup as part of call to g_hash_table_insert, then you only need to clean up direct_pinvoke string and direct_pinvoke_split when allocated.

Reply:
In the alternative, if we only called `g_strdup` within `g_hash_table_insert`, wouldn't there still be the possibility of `g_strdup` failing and returning `NULL`?
I can add a `g_free` for `direct_pinvoke`.
As for `library_name` and `entrypoint_name`, I was thinking that the ownership would be moved into the HashTable, which has the `g_free` `GDestroyNotify key_destroy_func`. Am I misunderstanding how this works?

Reply:
If g_strdup would fail it will only insert NULL into the table, and you have other code making sure not to use NULL entries in the hash table, then if g_strdup would fail, it would at least not crash the process, so I think you could pass it directly as the value to the hash table.

The hash table will free values that it owns, so if you can guarantee that each code path won't hit an error case before you transferred each string into hash tables its fine, but having functions like this tend to cause memory leaks and hard to maintain correctly without introducing memory leaks when changed, so having clear ownership is preferred and its simple as setting the variable to NULL when you transferred ownership and in you cleanup, do a g_free since that handles variables with NULL, that would make things more clear and simpler to maintain.

Reply:
After refactoring, `g_strdup` failing will error and propagate up to exit the mono aot compiler. We wouldn't want to proceed if it misses a specified direct pinvoke right?

Reply:
How does NativeAOT behave in such a case?

Reply:
OK, lets start defensive, failing the complete aot compile if we fail to handle the direct-pinvoke.